internal/primitives/state-machine/overlayedchanges/ext.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package overlayedchanges
+
+import (
+	"github.com/ChainSafe/gossamer/internal/primitives/state-machine/backend"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// storageAppend is a helper struct that appends a StorageValue to a slice.
+type storageAppend struct {
+	data *backend.StorageValue
+}
+
+// newStorageAppend creates a new storageAppend instance.
+func newStorageAppend(data *backend.StorageValue) *storageAppend {
+	return &storageAppend{data: data}
+}
+
+// extractLength extracts the scale encoded length of the StorageValue.
+func (sa *storageAppend) extractLength() *uint {
+	var length uint
+	err := scale.Unmarshal(*sa.data, &length)
+	if err != nil {
+		return nil
+	}
+	return &length
+}
+
+// replaceLength replaces the length of the StorageValue using scale encoding.
+func (sa *storageAppend) replaceLength(oldLength *uint, newLength uint) {
+	oldLenEncodedLen := 0
+	if oldLength != nil {
+		oldLenEncodedLen = compactLen(*oldLength)
+	}
+	newLenEncoded, _ := scale.Marshal(newLength)
+	data := spliceSlice(*sa.data, 0, oldLenEncodedLen, newLenEncoded)
+	newStorageValue := backend.StorageValue(data)
+	*sa.data = newStorageValue
+}
+
+// appendRaw appends a raw byte slice to the current StorageValue.
+func (sa *storageAppend) appendRaw(value []byte) {
+	*sa.data = append(*sa.data, value...)
+}
+
+// spliceSlice is a helper function that replaces a slice of elements with a new slice in the given positions.
+func spliceSlice[T any](slice []T, startIdx, endIdx int, replacement []T) []T {
+	if startIdx < 0 || endIdx > len(slice) || startIdx > endIdx {
```

Comment:
if the `endIdx` is eq to `len(slice) ` we can have problems right, given then `len(slice)` is not a valid index in the array, so should we change it to be `endIdx >= len(slice)`?
```suggestion
	if startIdx < 0 || endIdx >= len(slice) || startIdx > endIdx {
```